File: anagrams/ruby/bag.rb
require 'test/unit'

class Bag
  Primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101]

  def initialize (str)
    @product = 1
    str.downcase().each_byte {
      |b|
      if (b >= ?a and b <= ?z)
        @product *= Primes[b - ?a]
      end
    }
  end

  def hash
    @product.hash
  end
  
  def empty
    1 == @product
  end

  def ==(other)
    @product == other.product
  end

  def eql?(other)
    @product == other.product
  end

  def -(other)
    if(@product % other.product != 0)
      nil
    else
      p = Bag.new("")
      p.product = @product / other.product
      p
    end
  end

  def product
    @product
  end

  protected

  def product=(x)
    @product = x
  end
end

class Bag_Tests < Test::Unit::TestCase
  def test_joe_bob
    assert(Bag.new("").empty(), "initially empty")
    assert((not (Bag.new("a").empty())), "not always empty")
    assert_equal(Bag.new("abc"), Bag.new("cba"))
    assert((not(Bag.new("abc") == Bag.new("bc"))))
    assert_equal(Bag.new("a"),
                 Bag.new("ab") - Bag.new("b"))
    assert_equal(Bag.new("b"),
                 Bag.new("ab") - Bag.new("a"))
    assert((not( Bag.new("a") - Bag.new("b"))))
    assert((not( Bag.new("a") - Bag.new("aa"))))
    silly_long_string = "When first I was a wee, wee lad\n\
Eating on my horse\n\
I had to take a farting duck\n\
Much to my remorse.\n\
Oh Sally can't you hear my plea\n\
When Roe V Wade is nigh\n\
And candles slide the misty morn\n\
With dimples on your tie."

    ever_so_slightly_longer_string = silly_long_string + "x"
    assert_equal(Bag.new("x"),
                 Bag.new(ever_so_slightly_longer_string) - Bag.new(silly_long_string))
    
    assert_equal(Bag.new("abc"),
                 Bag.new("ABC"), "properly insensitive to case")
    
  end
end

File: anagrams/c++/bag.cc
#include <cassert>
#include <string>
#include <iostream>
#include <algorithm>

#include "bag.h"

bag::bag (const std::string &s)
{
  for (int i = 0;
       i < s.size ();
       i++)
    {
      if (isalpha (s.at (i)))
        _letters += (tolower (s.at(i)));
    }
  sort (_letters.begin (), _letters.end ());
}

bag*
bag::subtract_bag (const bag &subtrahend) const
{
  std::string b1(this->_letters);
  std::string b2 (subtrahend._letters);

  std::string difference;
  for (;;
       b1 = b1.substr (1), b2 = b2.substr (1)) 
    {
      if (!b2.size ())
        return new bag (difference + b1);

      if (!b1.size ())
        return 0;

      if (b1.at (0) != b2.at (0))
        {
          char c1 = b1.at(0);
          char c2 = b2.at(0);

          while ((c1 < c2))
            {
              if (!b1.size ())
                return 0;

              difference += c1;
              
              b1 = b1.substr (1);
              if (!b1.size ())
                return 0;

              c1 = b1.at (0);
            }

          if (c1 > c2)
            return 0;
        }  
    }
  bag *rv = new bag (difference);
  assert (rv->size () == this->size () - subtrahend.size ());
  return rv;
}
#if 0
int
main ()
{
#if 0
  assert (bag ("HEY") == bag ("hey"));

  assert (bag ("").is_empty ());


  assert (!bag ("a").is_empty ());

  assert (bag ("abc") == bag ("cba")) ;


  assert (!(bag ("abc") == bag ("bc")));
#endif
  assert (!(bag ("abc") == bag ("a")));

  {
    bag *oughta_be_empty (bag ("a").subtract_bag (bag ("a")));
    assert (oughta_be_empty);
    assert (oughta_be_empty->is_empty ());
  }

  bag *b = 0;

  // oh how I miss garbage collection.
  b = bag("ab").subtract_bag (bag ("b"));
  assert (bag ("a") == *b) ;
  delete b; b = 0;

  assert (!bag ("a").subtract_bag (bag ("b")));
  assert (!bag ("a").subtract_bag (bag ("aa")));


  const std::string silly_long_string = "When first I was a wee, wee lad Eating on my horse I had to take a farting duck Much to my remorse. Oh Sally can't you hear my plea When Roe V Wade is nigh And candles slide the misty morn With dimples on your tie.";

  const std::string ever_so_slightly_longer_string (silly_long_string + "x");

  b = bag (ever_so_slightly_longer_string).subtract_bag (bag (silly_long_string));
  assert (*b == bag ("x"));
  delete b; b = 0;

  std::cout << "We cool!\n";
}
#endif
